src/mantarray_file_manager/file_writer.py

```diff
@@ -156,3 +165,150 @@ def migrate_to_latest_version(
+        from_end: centimilliseconds to trim from the end
+
+    Returns:
+        The path to the trimmed H5 file. The amount actually trimmed off the file is dependent on the timepoints of the tissue sensor data and will be reflected in the new file name, message to the terminal, and the metadata. If the amount to be trimmed off is in between two time points, less time will be trimmed off and the lower timepoint will be used if from_start or upper timepoint if from_last. Reference sensor readings are trimmed according to the amount trimmed from tissue data.
+    """
+    # pylint: disable-msg=too-many-locals
+    validate_int(
+        value=from_start,
+        allow_null=True,
+        minimum=0,
+    )
+    validate_int(value=from_end, allow_null=True, minimum=0)
+
+    if not from_start and not from_end:
+        raise UnsupportedArgumentError()
+
+    old_file_version = _get_format_version_of_file(file_path)
+
+    if old_file_version != CURRENT_HDF5_FILE_FORMAT_VERSION:
+        raise UnsupportedMantarrayFileVersionForTrimmingError(old_file_version)
+
+    working_directory = getcwd()
+    old_file = WellFile(file_path)
+    old_file_basename = ntpath.basename(file_path)[:-3]
+
+    # old metadata
+    old_h5_file = old_file.get_h5_file()
+    old_metadata_keys = set(old_h5_file.attrs.keys())
+    old_from_end = 0
+    old_from_start = 0
+
+    if str(IS_FILE_ORIGINAL_UNTRIMMED_UUID) in old_metadata_keys:
+        old_from_start = old_h5_file.attrs[str(TRIMMED_TIME_FROM_ORIGINAL_START_UUID)]
+        old_from_end = old_h5_file.attrs[str(TRIMMED_TIME_FROM_ORIGINAL_END_UUID)]
+
+        old_metadata_keys.remove(str(TRIMMED_TIME_FROM_ORIGINAL_START_UUID))
+        old_metadata_keys.remove(str(TRIMMED_TIME_FROM_ORIGINAL_END_UUID))
+        old_metadata_keys.remove(str(IS_FILE_ORIGINAL_UNTRIMMED_UUID))
+
+        old_file_basename = old_file_basename.split("__trimmed")[0]
+
+    # finding amount to trim
+    old_raw_reference_data = old_file.get_raw_reference_reading()
+    old_tissue_data = old_file.get_raw_tissue_reading()
+
+    tissue_data_start_val = old_tissue_data[0][0]
+    tissue_data_last_val = old_tissue_data[0][-1]
+    tissue_data_start_index = _find_start_index(from_start, old_tissue_data[0])
+    tissue_data_last_index = _find_last_index(from_end, old_tissue_data)
+
+    actual_start_trimmed = (
+        old_tissue_data[0][tissue_data_start_index] - tissue_data_start_val
+    )
+    actual_end_trimmed = (
+        tissue_data_last_val - old_tissue_data[0][tissue_data_last_index]
+    )
+
+    reference_data_start_index = _find_start_index(
+        actual_start_trimmed, old_raw_reference_data[0]
+    )
+    reference_data_last_index = _find_last_index(
+        actual_end_trimmed, old_raw_reference_data
+    )
+
+    if (
+        reference_data_start_index >= reference_data_last_index
+        or tissue_data_start_index >= tissue_data_last_index
+    ):
+        raise TooTrimmedError(from_start, from_end)
```

Comment:
in general, a fundamental error related to function input arguments should be identified as early as possible. for example prior to removing old metadata keys